### PR TITLE
[ADDED] UserInfoHandler for dynamically setting user/password

### LINKED ISSUE #1694: Explicit support for auth change during reconnect
### Proposed change

I am experiencing "fuzzy" behavior regarding the following. Fuzzy as in, the code works when the NATS client is connected to a single URL, but it doesnt work when connected to multiple URLs. 

We are using the NATS Auth Callout, but client side, the Password field needs to change each time the disconnect callback is called, which happens when the auth callout JWTs expire (eg hourly). Specifically, our autth callout backend is interpreting a "password" specific to our in house IAM stack, and those *also* expire. 

Because our auth callout tokens expire every hour, our `disconnectedCB` is called every hour. And in that, we want to change `conn.Opts.Password` so that it uses the new password (the client has to regenerate one at this time) on the reconnection attempt (see below, the actual auth call is made after the reconnected callback, so "hook wise", we already have a hook at the right spot.)

My functions currently looks like this:

```go
func (gb *Base) tokenAndConnect(ctx context.Context) (*nats.Conn, error) {
	iamCreds, err := gb.iamRuntime.GetAccessToken(ctx, &identity.GetAccessTokenRequest{})
	if err != nil {
		gb.Logger.Fatalf("failed to get access token: %v", err)
	}

	// For the auth callout, we have a sentinel and we also need an Identity token as our password
	return nats.Connect(
		gb.config.Nats.URL,
		nats.UserCredentials(gb.config.Nats.SentinelCredsFile),
		nats.UserInfo(gb.config.MsName, iamCreds.Token),
	)
}


func (gb *Base) AddNATSConn(ctx context.Context) *Base {
	nc, err := gb.tokenAndConnect(ctx)
	if err != nil {
		gb.Logger.Fatalf("failed to create Initial NATS connection: %v", err)
	}

	/*
		    Disconnect Callbacks are called when the function reconnects....
			https://github.com/nats-io/nats.go/blob/370bc4fc4cac2de8d5dbbddbb98739f9cb2d7f8b/nats.go#L2779-L2787
			The purpose of this is to get a new identityJWT before the reconnect happens

			We cant use the reconnectCB (misleading name here)
			ReconnectCB is called *after* the reconnection happens. In order to refresh the token before the reconnect,
			Doc strings for variouis callbacks:
			ReconnectedCB sets the reconnected handler called whenever the connection is successfully reconnected.
			ConnectedCB sets the connected handler called when the initial connection is established. It is not invoked on successful reconnects - for reconnections, use ReconnectedCB
			DisconnectedErrCB sets the disconnected error handler that is called whenever the connection is disconnected.
			ClosedCB sets the closed handler that is called when a client will no longer be connected.
	*/

	nc.SetDisconnectErrHandler(func(conn *nats.Conn, err error) {
		gb.Logger.Infof("NATS connection disconnected: %v. Generating new Identity JWT", err)
		iamCreds, err := gb.iamRuntime.GetAccessToken(ctx, &identity.GetAccessTokenRequest{})
		if err != nil {
			gb.Logger.Fatalf("Failed to get access token: %w", err)
		}
		
                 // *********************************!!!
                 // doesnt work in multi URL case but works fine when you seed the client with a single NATS server
		conn.Opts.Password = iamCreds.Token
	})

	nc.SetReconnectHandler(func(conn *nats.Conn) {
		gb.Logger.Infof("NATS connection reconnected. Reconnect count %d", conn.Reconnects)
	})

	gb.NatsConn = nc
	return gb
}
```

I think it has something to do with this warning which I just saw..
https://github.com/nats-io/nats.go/blob/e3df53d7fa82d1d60335c53b66b43d9be90a8881/nats.go#L537-L538

It works in the case where the client only was given one cluster URL but not multiple. 

I eventually tracked down this call chain on reconnects:
1. the disconnect handler is called first: https://github.com/nats-io/nats.go/blob/e3df53d7fa82d1d60335c53b66b43d9be90a8881/nats.go#L2781-L2787
2. later in that function, createConn is called though I dont think this is the problematic one: https://github.com/nats-io/nats.go/blob/e3df53d7fa82d1d60335c53b66b43d9be90a8881/nats.go#L2873 
3. then later down in that function, nc.processConnectInit() is called: https://github.com/nats-io/nats.go/blob/e3df53d7fa82d1d60335c53b66b43d9be90a8881/nats.go#L2886
4. that calls sendConnect(): https://github.com/nats-io/nats.go/blob/e3df53d7fa82d1d60335c53b66b43d9be90a8881/nats.go#L2380
5. that calls connectProto https://github.com/nats-io/nats.go/blob/e3df53d7fa82d1d60335c53b66b43d9be90a8881/nats.go#L2644
6. that contains: https://github.com/nats-io/nats.go/blob/e3df53d7fa82d1d60335c53b66b43d9be90a8881/nats.go#L2551-L2566

that code makes it unclear when this `Password` change is going to be reflected. 

My fear/suspicion is that, in the multiple URL case, that `User` is populated; triggering the `if`, which is why it sends the old token, which is why it breaks. In the single URL case, that is not populated, thus triggering the `else`

I am trying to see how I can explicitly force  it to use a new password, either via setting some variable sure to be picked up, or via a callback. Maybe a flag that says "always override the auth from the connection options". 

This block might be related, I was trying to find anywhere where `nc.current.url` is set: https://github.com/nats-io/nats.go/blob/e3df53d7fa82d1d60335c53b66b43d9be90a8881/nats.go#L1821-L1827

### Use case

We are using the NATS Auth Callout, but client side, the Password field needs to change each time the disconnect callback is called, which happens when the auth callout JWTs expire (eg hourly). Specifically, our autth callout backend is interpreting a "password" specific to our in house IAM stack.

### Contribution

_No response_

### PR DESCRIPTION
**Pull request by piotrpio**
Resolves #1694

Signed-off-by: Piotr Piotrowski <piotr@synadia.com>

### CODE REVIEW

File: nats.go [hunk 1]
```diff
@@ -1166,6 +1173,13 @@ func UserInfo(user, password string) Option {
 	}
 }
 
+func UserInfoHandler(cb UserInfoCB) Option {
```
Comment (Jarema) [inline nats.go:1176]: the name `UserInfo` is a bit ambigous. 
Maybe something hinting it's about Username and password being set?

**Review by Jarema — CHANGES_REQUESTED** [review]
Looks ok, but I don't think it's a good name.

File: nats.go [hunk 2]
```diff
@@ -2563,6 +2577,13 @@ func (nc *Conn) connectProto() (string, error) {
 		pass = o.Password
 		token = o.Token
 		nkey = o.Nkey
+
+		if nc.Opts.UserInfo != nil {
```
Comment (tommyjcarpenter) [inline nats.go:2581]: ...this tiny change will have huge opportunities for our organization (from #1694)!

File: nats.go (hunk 1 above)
Comment (Jarema) [reply nats.go:1176]: Ok, understood that its done for consistency.

**Review by Jarema — APPROVED** [review]
LGTM

**Comment by tommyjcarpenter** [comment]
Is there a setter for this like some of the other CBs?

**Pull request merged**